File: dkplayer-ui/src/main/java/xyz/doikki/videocontroller/component/TitleView.java
package xyz.doikki.videocontroller.component;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import xyz.doikki.videocontroller.R;
import xyz.doikki.videoplayer.VideoView;
import xyz.doikki.videoplayer.controller.component.ControlComponent;
import xyz.doikki.videoplayer.render.ScreenMode;
import xyz.doikki.videoplayer.util.PlayerUtils;

/**
 * 播放器顶部标题栏
 */
public class TitleView extends BaseControlComponent implements ControlComponent {

    private LinearLayout mTitleContainer;
    private TextView mTitle;
    private TextView mSysTime;//系统当前时间

    private BatteryReceiver mBatteryReceiver;

    private boolean mIsRegister;//是否注册BatteryReceiver

    public TitleView(@NonNull Context context) {
        super(context);
    }

    public TitleView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TitleView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void setupViews() {
        setVisibility(GONE);
        LayoutInflater.from(getContext()).inflate(R.layout.dkplayer_layout_title_view, this, true);
        mTitleContainer = findViewById(R.id.title_container);
        ImageView back = findViewById(R.id.back);
        mTitle = findViewById(R.id.title);
        mSysTime = findViewById(R.id.sys_time);
        //电量
        ImageView batteryLevel = findViewById(R.id.iv_battery);

        if (isFocusUiMode()) {
            //tv模式不要电量，不要返回按钮
            back.setVisibility(View.GONE);
            batteryLevel.setVisibility(View.GONE);

            //因为返回按钮不可见，因此标题左侧设置多一个margin
            ViewGroup.LayoutParams titleLp = mTitle.getLayoutParams();
            if (titleLp instanceof MarginLayoutParams) {
                MarginLayoutParams lp = (MarginLayoutParams) titleLp;
                lp.setMargins(lp.leftMargin + getContext().getResources().getDimensionPixelSize(R.dimen.dkplayer_controller_icon_padding),
                        lp.topMargin, lp.rightMargin, lp.bottomMargin);
                mTitle.setLayoutParams(lp);
            }
        } else {
            back.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    Activity activity = PlayerUtils.scanForActivity(getContext());
                    if (activity != null && mControlWrapper.isFullScreen()) {
                        mControlWrapper.stopFullScreen();
                    }
                }
            });
            mBatteryReceiver = new BatteryReceiver(batteryLevel);
        }
    }

    public void setTitle(String title) {
        mTitle.setText(title);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (mIsRegister && mBatteryReceiver != null) {
            getContext().unregisterReceiver(mBatteryReceiver);
            mIsRegister = false;
        }
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (!mIsRegister && mBatteryReceiver != null) {
            getContext().registerReceiver(mBatteryReceiver, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
            mIsRegister = true;
        }
    }


    @Override
    public void onVisibilityChanged(boolean isVisible, Animation anim) {
        //只在全屏时才有效
        if (!mControlWrapper.isFullScreen()) return;
        if (isVisible) {
            if (getVisibility() == GONE) {
                mSysTime.setText(PlayerUtils.getCurrentSystemTime());
                setVisibility(VISIBLE);
                if (anim != null) {
                    startAnimation(anim);
                }
            }
        } else {
            if (getVisibility() == VISIBLE) {
                setVisibility(GONE);
                if (anim != null) {
                    startAnimation(anim);
                }
            }
        }
    }

    @Override
    public void onPlayStateChanged(int playState) {
        switch (playState) {
            case VideoView.STATE_IDLE:
            case VideoView.STATE_START_ABORT:
            case VideoView.STATE_PREPARING:
            case VideoView.STATE_PREPARED:
            case VideoView.STATE_ERROR:
            case VideoView.STATE_PLAYBACK_COMPLETED:
                setVisibility(GONE);
                break;
        }
    }

    @Override
    public void onScreenModeChanged(int screenMode) {
        if (screenMode == ScreenMode.FULL) {
            if (mControlWrapper.isShowing() && !mControlWrapper.isLocked()) {
                setVisibility(VISIBLE);
                mSysTime.setText(PlayerUtils.getCurrentSystemTime());
            }
            mTitle.setSelected(true);
        } else {
            setVisibility(GONE);
            mTitle.setSelected(false);
        }

        Activity activity = PlayerUtils.scanForActivity(getContext());
        if (activity != null && mControlWrapper.hasCutout()) {
            int orientation = activity.getRequestedOrientation();
            int cutoutHeight = mControlWrapper.getCutoutHeight();
            if (orientation == ActivityInfo.SCREEN_ORIENTATION_PORTRAIT) {
                mTitleContainer.setPadding(0, 0, 0, 0);
            } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE) {
                mTitleContainer.setPadding(cutoutHeight, 0, 0, 0);
            } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE) {
                mTitleContainer.setPadding(0, 0, cutoutHeight, 0);
            }
        }
    }

    @Override
    public void onLockStateChanged(boolean isLocked) {
        if (isLocked) {
            setVisibility(GONE);
        } else {
            setVisibility(VISIBLE);
            mSysTime.setText(PlayerUtils.getCurrentSystemTime());
        }
    }

    private static class BatteryReceiver extends BroadcastReceiver {
        private final ImageView pow;

        public BatteryReceiver(ImageView pow) {
            this.pow = pow;
        }

        @Override
        public void onReceive(Context context, Intent intent) {
            Bundle extras = intent.getExtras();
            if (extras == null) return;
            int current = extras.getInt("level");// 获得当前电量
            int total = extras.getInt("scale");// 获得总电量
            int percent = current * 100 / total;
            pow.getDrawable().setLevel(percent);
        }
    }
}

File: dkplayer-ui/src/main/java/xyz/doikki/videocontroller/component/TitleView.kt
package xyz.doikki.videocontroller.component

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ActivityInfo
import android.util.AttributeSet
import android.view.animation.Animation
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import xyz.doikki.videocontroller.R
import xyz.doikki.videoplayer.VideoView
import xyz.doikki.videoplayer.util.PlayerUtils
import xyz.doikki.videoplayer.util.orDefault

/**
 * 播放器顶部标题栏
 */
class TitleView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : BaseControlComponent(context, attrs, defStyleAttr) {

    private val titleContainer: LinearLayout
    private val titleView: TextView
    private val sysTime: TextView//系统当前时间

    private lateinit var batteryReceiver: BatteryReceiver

    //是否注册BatteryReceiver
    private var batteryReceiverRegistered = false

    /**
     * 是否启用电量检测功能
     */
    private var batteryEnabled: Boolean = true

    fun setTitle(title: String?) {
        titleView.text = title
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (batteryEnabled && batteryReceiverRegistered) {
            context.unregisterReceiver(batteryReceiver)
            batteryReceiverRegistered = false
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (batteryEnabled && !batteryReceiverRegistered) {
            context.registerReceiver(batteryReceiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            batteryReceiverRegistered = true
        }
    }

    override fun onVisibilityChanged(isVisible: Boolean, anim: Animation?) {
        //只在全屏时才有效
        if (!controller?.isFullScreen.orDefault()) return
        if (isVisible) {
            if (visibility == GONE) {
                sysTime.text = PlayerUtils.getCurrentSystemTime()
                visibility = VISIBLE
                anim?.let { startAnimation(it) }
            }
        } else {
            if (visibility == VISIBLE) {
                visibility = GONE
                anim?.let { startAnimation(it) }
            }
        }
    }

    override fun onPlayStateChanged(playState: Int, extras: HashMap<String, Any>) {
        when (playState) {
            VideoView.STATE_IDLE, VideoView.STATE_START_ABORT,
            VideoView.STATE_PREPARING, VideoView.STATE_PREPARED,
            VideoView.STATE_ERROR, VideoView.STATE_PLAYBACK_COMPLETED -> visibility = GONE
        }
    }

    @SuppressLint("SwitchIntDef")
    override fun onScreenModeChanged(screenMode: Int) {
        val controller = this.controller
        if (screenMode == VideoView.SCREEN_MODE_FULL) {
            if (controller != null && controller.isShowing && !controller.isLocked) {
                visibility = VISIBLE
                sysTime.text = PlayerUtils.getCurrentSystemTime()
            }
            titleView.isSelected = true
        } else {
            visibility = GONE
            titleView.isSelected = false
        }
        val activity = this.activity
        if (activity != null && controller != null && controller.hasCutout == true) {
            val orientation = activity.requestedOrientation
            val cutoutHeight = controller.cutoutHeight
            when (orientation) {
                ActivityInfo.SCREEN_ORIENTATION_PORTRAIT -> {
                    titleContainer.setPadding(0, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE -> {
                    titleContainer.setPadding(cutoutHeight, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE -> {
                    titleContainer.setPadding(0, 0, cutoutHeight, 0)
                }
            }
        }
    }

    override fun onLockStateChanged(isLocked: Boolean) {
        if (isLocked) {
            visibility = GONE
        } else {
            visibility = VISIBLE
            sysTime.text = PlayerUtils.getCurrentSystemTime()
        }
    }

    private class BatteryReceiver(private val pow: ImageView) : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val extras = intent.extras ?: return
            val current = extras.getInt("level") // 获得当前电量
            val total = extras.getInt("scale") // 获得总电量
            val percent = current * 100 / total
            pow.drawable.level = percent
        }
    }

    init {
        visibility = GONE
        if (isTelevisionUiMode()) {
            batteryEnabled = false
            layoutInflater.inflate(R.layout.dkplayer_layout_title_view_tv, this)
            //tv模式不要电量，不要返回按钮
            findViewById<ImageView>(R.id.back)?.visibility = GONE
            findViewById<ImageView>(R.id.iv_battery)?.visibility = GONE
        } else {
            batteryEnabled = true
            layoutInflater.inflate(R.layout.dkplayer_layout_title_view, this)
            findViewById<ImageView>(R.id.back).setOnClickListener {
                val activity = activity
                if (activity != null && controller?.isFullScreen.orDefault()) {
                    controller?.stopFullScreen()
                }
            }
            //电量
            val batteryLevel = findViewById<ImageView>(R.id.iv_battery)
            batteryReceiver = BatteryReceiver(batteryLevel)
        }
        titleContainer = findViewById(R.id.title_container)
        titleView = findViewById(R.id.title)
        sysTime = findViewById(R.id.sys_time)
    }
}